### ISSUE TITLE
exp run: AssertionError: self.is_in_repo with cached outputs

### ISSUE THREAD

**Issue by asiron** [iterative/dvc#9615]
# Bug Report

## Description

I found a difficult to reproduce bug. It seems to have something to do with caching the output of a stage
Here is the `dvc.yaml`. When `cache: true` (default option) on a non-deterministic output (could be a model for example, if training is not fully deterministic) then I get an AssertionError. Setting `cache: false` fixes the problem.

```yaml
stages:
  preprocess:
    cmd: python src/preprocess.py
    deps:
    - data/raw
    - src/preprocess.py
    params:
    - preprocessing
    - unprocessed_data
    outs:
    - ${preprocessing.output_path}
  test:
    cmd: 'echo "(${preprocessing.resize_to.width},${preprocessing.resize_to.height})" > data/result.txt'
    deps:
    - ${preprocessing.output_path}
    outs:
    - data/result.txt:
        cache: true
```

. The following commands fail: `dvc exp run`, `dvc repro`, `dvc data status`.  Here is the stack trace: 

```console
(.env) ➜  magic_mvp git:(dev) ✗ dvc data status -v
2023-06-16 12:44:45,947 DEBUG: v3.0.0 (pip), CPython 3.10.6 on Linux-5.19.0-43-generic-x86_64-with-glibc2.35
2023-06-16 12:44:45,947 DEBUG: command: /home/mzurad/code/magic_mvp/.env/bin/dvc data status -v
2023-06-16 12:44:46,161 ERROR: unexpected error
Traceback (most recent call last):
  File "/home/mzurad/code/magic_mvp/.env/lib/python3.10/site-packages/dvc/cli/__init__.py", line 209, in main
    ret = cmd.do_run()
  File "/home/mzurad/code/magic_mvp/.env/lib/python3.10/site-packages/dvc/cli/command.py", line 26, in do_run
    return self.run()
  File "/home/mzurad/code/magic_mvp/.env/lib/python3.10/site-packages/dvc/commands/data.py", line 114, in run
    status = self.repo.data_status(
  File "/home/mzurad/code/magic_mvp/.env/lib/python3.10/site-packages/dvc/repo/data.py", line 215, in status
    uncommitted_diff = _diff_index_to_wtree(
  File "/home/mzurad/code/magic_mvp/.env/lib/python3.10/site-packages/dvc/repo/data.py", line 153, in _diff_index_to_wtree
    workspace = build_data_index(
  File "/home/mzurad/code/magic_mvp/.env/lib/python3.10/site-packages/dvc/repo/index.py", line 663, in build_data_index
    hash_name = _get_entry_hash_name(index, workspace, key)
  File "/home/mzurad/code/magic_mvp/.env/lib/python3.10/site-packages/dvc/repo/index.py", line 724, in _get_entry_hash_name
    src_entry = index.data[workspace][key]
  File "/home/mzurad/code/magic_mvp/.env/lib/python3.10/site-packages/funcy/objects.py", line 25, in __get__
    res = instance.__dict__[self.fget.__name__] = self.fget(instance)
  File "/home/mzurad/code/magic_mvp/.env/lib/python3.10/site-packages/dvc/repo/index.py", line 421, in data
    _load_storage_from_out(data_index.storage_map, key, out)
  File "/home/mzurad/code/magic_mvp/.env/lib/python3.10/site-packages/dvc/repo/index.py", line 163, in _load_storage_from_out
    storage_map.add_cache(ObjectStorage(key, out.cache))
  File "/home/mzurad/code/magic_mvp/.env/lib/python3.10/site-packages/dvc/output.py", line 528, in cache
    assert self.is_in_repo
AssertionError

2023-06-16 12:44:46,174 DEBUG: link type reflink is not available ([Errno 95] no more link types left to try out)
2023-06-16 12:44:46,174 DEBUG: Removing '/home/mzurad/code/.YYn83na5FXpT7qwNbiv5Mt.tmp'
2023-06-16 12:44:46,174 DEBUG: link type hardlink is not available ([Errno 95] no more link types left to try out)
2023-06-16 12:44:46,174 DEBUG: Removing '/home/mzurad/code/.YYn83na5FXpT7qwNbiv5Mt.tmp'
2023-06-16 12:44:46,174 DEBUG: Removing '/home/mzurad/code/.YYn83na5FXpT7qwNbiv5Mt.tmp'
2023-06-16 12:44:46,174 DEBUG: Removing '/mnt/crucial/.dvc-cache/files/md5/.Coc9MgNETDmKtwiCqS7Ugz.tmp'
```

### Reproduce

I can't seem to reproduce this bug, I tried with the following but this works. My guess is the issue might arise when data is cached or sent to the remote.

```bash
cd bug

python3 -m venv .env
source .env/bin/activate
pip install dvc==3.0.0

git init .
dvc init 
dvc config hydra.enabled True


echo "123" > a.txt

mkdir -p conf
echo "defaults:\n- test" > conf/config.yaml
echo "some_param: intermediate.txt" > conf/test.yaml

git add .
git commit -m 'test'

cat > dvc.yaml <<EOL
stages:
  first:
    cmd: cat a.txt > \${some_param}
    deps:
    - a.txt
    params:
    - some_param
    outs:
    - \${some_param}
  second:
    cmd: cat /dev/urandom | head -n1 | cut -c-6 > c.txt
    deps:
    - \${some_param}
    params:
    - some_param
    outs:
    - c.txt

EOL

dvc exp run 

```

### Expected

`dvc exp run`, `dvc repro` and `dvc data status` should work correctly.

### Environment information

**Output of `dvc doctor`:**

```console
DVC version: 3.0.0 (pip)
------------------------
Platform: Python 3.10.6 on Linux-5.19.0-43-generic-x86_64-with-glibc2.35
Subprojects:
        dvc_data = 1.11.0
        dvc_objects = 0.23.0
        dvc_render = 0.5.3
        dvc_task = 0.3.0
        scmrepo = 1.0.3
Supports:
        http (aiohttp = 3.8.4, aiohttp-retry = 2.8.3),
        https (aiohttp = 3.8.4, aiohttp-retry = 2.8.3)
Config:
        Global: /home/mzurad/.config/dvc
        System: /etc/xdg/xdg-ubuntu/dvc
Cache types: symlink
Cache directory: ext4 on /dev/mapper/cryptpart
Caches: local
Remotes: local
Workspace directory: ext4 on /dev/mapper/ubuntu--vg-root
Repo: dvc, git
Repo.site_cache_dir: /var/tmp/dvc/repo/53229a3220cd5e3d302d366eaae0fb54

```

**Additional Information (if any):**
Not sure if relevant but `data/raw` is `dvc import`ed and here is the .dvc file:

```yaml
md5: b78019a3d2b4006faef6d1f7af27c1aa
frozen: true
deps:
- path: my_directory
  repo:
    url: git@gitlab.com:some_user/some_repo.git
    rev: some-branch
    rev_lock: 4f038c69efecebd4c238a61767598d1cffb5ea84
outs:
- md5: e35a908611236136f99af7da7c5c57fc.dir
  size: 330932201
  nfiles: 109
  hash: md5
  path: raw
```

and here is my local `config` file:

```yaml
[core]
    analytics = false
    remote = local
[cache]
    type = symlink
    dir = /mnt/crucial/.dvc-cache
[hydra]
    enabled = True
    config_dir = config
['remote "local"']
    url = /home/mzurad/.dvc-store
```

**Comment by dberenbaum** [comment]
@asiron Could you try with `-vv` and post the output?

**Comment by efiop** [comment]
This should be fixed in 3.1.0 , please give a try and let us know if that works.
